api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,309 @@
+			utils.Logger().Info().
+				Msgf("[SYNC] Node is now IN SYNC! (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 60
+		}
+
+		utils.Logger().Info().
+			Msgf("[SYNC] Node is OUT OF SYNC (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+				isBeacon, bc.ShardID(), maxHeight, height)
+
+		var heights []uint64
+		curEpoch := block.Epoch()
+		for i := 0; i < int(SyncLoopBatchSize); i++ {
+			epochLastBlock := shard.Schedule.EpochLastBlock(curEpoch.Uint64())
+			if epochLastBlock > maxHeight {
+				break
+			}
+			if height != epochLastBlock {
+				heights = append(heights, epochLastBlock)
+			}
+			curEpoch = curEpoch.Add(curEpoch, common.Big1)
+		}
+
+		if len(heights) == 0 {
+			// We currently on top.
+			return 10
+		}
+
+		err := ss.ProcessStateSync(heights, bc, worker)
+		if err != nil {
+			utils.Logger().Error().Err(err).
+				Msgf("[SYNC] ProcessStateSync failed (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 2
+		}
+	}
+}
+
+// ProcessStateSync processes state sync from the blocks received but not yet processed so far
+func (ss *EpochSync) ProcessStateSync(heights []uint64, bc *core.BlockChain, worker *worker.Worker) error {
+
+	var payload [][]byte
+
+	ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
+		resp := peerConfig.GetClient().GetBlocksByHeights(heights)
+		if resp == nil {
+			return false
+		}
+		if len(resp.Payload) == 0 {
+			return false
+		}
+		payload = resp.Payload
+		return true
+	})
+
+	if len(payload) == 0 {
+		return errors.New("empty payload")
+	}
+
+	decoded := make([]*types.Block, 0, len(payload))
+	for idx, blockBytes := range payload {
+		block, err := RlpDecodeBlockOrBlockWithSig(blockBytes)
+		if err != nil {
+			return errors.Wrap(err, "failed decode")
+		}
+
+		if !block.IsLastBlockInEpoch() {
+			return errors.Errorf("block is not last index %d hash %s", idx, block.Hash().Hex())
+		}
+
+		decoded = append(decoded, block)
+	}
+
+	for _, block := range decoded {
+		sig, bitmap, err := chain.ParseCommitSigAndBitmap(block.GetCurrentCommitSig())
+		if err != nil {
+			return errors.Wrap(err, "parse commitSigAndBitmap")
+		}
+
+		// Signature validation.
+		err = bc.Engine().VerifyHeaderSignature(bc, block.Header(), sig, bitmap)
+		if err != nil {
+			return errors.Wrap(err, "failed signature validation")
+		}
+
+		ep := block.Header().Epoch()
+		next := ep.Add(ep, common.Big1)
+		_, err = bc.StoreShardStateBytes(next, block.Header().ShardState())
+		if err != nil {
+			return err
+		}
+
+		err = bc.WriteBlockWithoutState(block, common.Big1)
+		if err != nil {
+			return err
+		}
+
+		err = bc.WriteHeadBlock(block)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getConsensusHashes gets all hashes needed to download.
+func (ss *EpochSync) getConsensusHashes(startHash []byte, size uint32) error {
```

Comment:
this and CreateSyncConfig method are all copy/paste, please use inheritance to refactor them so it's reused.

Reply:
fixed